### ISSUE TITLE
Support WASM to allow people to use native zmq in the web browser

### ISSUE THREAD

**Issue by TheButlah** [zeromq/zmq.rs#64]
One of Rust's strengths is is wasm support, letting developers run things natively in the web browser. It would be great if we could plan to support wasm in the future.

Looking into this, tokio currently [doesn't support wasm](https://github.com/tokio-rs/tokio/issues/1597), and async-std has a [minimal feature set that supports wasm](https://github.com/async-rs/async-std/issues/220). It seems like a trivial way to use tcp in the browser is not yet feasible, but we should be on the lookout for ways to support it in the near future. One possible approach might be to use websockets instead of tokio sockets, and perhaps unbake the assumption that the sockets are managed by tokio by abstracting them with a common trait.

**Comment by TheButlah** [comment]
I've added `ws://` and `wss://` to the transport todo list here: #65. Unsure if thats the right approach, but we can figure it out in the future.

**Comment by TheButlah** [comment]
Update: I've learned a bit more about web sockets vs tcp. Web sockets is a protocol, not a transport (i.e. HTTP vs tcp). It is built on top of TCP. However, you cannot connect a websocket client to a tcp socket unless the server is also using websockets - just like you can't make an HTTP request to a tcp socket unless its actually a HTTP server listening on that socket.

This is why the `ws://` and `wss://` transports exist - they descibe how ZMQ sockets work over web sockets. Therefore, you will never be able to do `tcp://` in the browser, but you can do `wss://` in the browser and connect to a server using `wss:://`.

**Comment by TheButlah** [comment]
I've looked into the ecosystem to support this a bit.

First, there is the question of the async executor. Tokio does not support WASM, and doesn't seem to have any plans to do so. async-std currently has paritial support for WASM - importantly, it builds for the platform and task::spawn works (according to [here](https://github.com/async-rs/async-std/issues/220)). Its important to note that we technically care only about the *executor*, not the other functionality of tokio/async-std, but we may consider transitioning as an option. We should try to take care to write our code as runtime-independent as possible.

Next, there is the question of how to get a websocket library that works with the async ecosystem. The best option looks to be [async-tungstenite)(https://github.com/sdroege/async-tungstenite).

**Comment by TheButlah** [comment]
@Alexei-Kornienko would you be open to a PR to get the code to work on both async-std and tokio, via a feature flag?

**Comment by Alexei-Kornienko** [comment]
@TheButlah In theory I'm open for such contributions. However I'm not sure that it will be easy to implement. You can take a look at codec code. We'll have to implement something like Codec and Framed for async-std

**Comment by TheButlah** [comment]
@Alexei-Kornienko looking into it, the Codec and Framed stuff shouldn't really depend on any particular async runtime - it just needs common traits from `futures` like `Stream` and `Sink`. Tokio provides them from `tokio-util`, but no alternative appears to exist in `async-std`, presumably because Codecs aren't a concept in the rust standard library. That being said, conceptually the functionality is quite simple.

I found https://crates.io/crates/futures_codec, which looks to have pretty low development velocity, but the crate is a pretty small amount of code and there isn't exactly a large featureset that we are relying on here. I think it ought to be feasible to use that crate in lieu of `tokio-utils`'s `Framed`.

Once we do that, I'm sure we can abstract over tokio and async-stds other concepts and try to rely exclusively on `futures` whenever possible for common functionality like traits and channels.

**Comment by Alexei-Kornienko** [comment]
@TheButlah I don't think that completely replacing a tokio code is a good idea. Instead I think it can hidden behind a feature flag. So for tokio variant it will compile with tokio provided codec implementation and for async-std it may have a separate codec impl.

**Comment by TheButlah** [comment]
Oh, I didn't mean to imply that we should replace tokio code across the board - just that we should strive to use runtime-agnostic crates when we can to ease transitions.

Most of the tokio-specific code seems to be in the `codec` module, as well as when binding/listening on the tcp transport. Since WASM doesn't have tcp, this code will be gated on a feature flag, and doesn't need to be ported to be runtime-agnostic (yet). We can use `async-tungstenite` to handle the websocket stuff, which should be runtime independent already. Then all thats left is the possibility that we will have some `spawn_blocking` or `block_on` calls which are runtime specific scattered in non-transport related code. In those cases, we will have to see what the best way to handle them is.

But overall, I think the path forward is probably the following:
1. Update the `codec` module to use `futures_codec` instead of `tokio-utils` (since that is common functionality that cant be feature-gated and would need to be runtime-agnostic)
2. Feature gate the TCP transport - it can continue to use regular tokio since it wont need to compile on WASM anyway
3. Find any other common code that isn't tcp specific that uses tokio-only features and see if there is anything tricky we need to do, or if making it runtime-agnostic is trivial.
4. Get the code to compile on WASM and add a CI check for that
5. Implement the ZWS 2.0 RFC

This should have a minimal impact on code that is heavily dependent on tokio, as most of that code is transport-specific and doesn't need to compile to WASM anyway

**Comment by TheButlah** [comment]
I've implemented some first steps for this in #77, it extracts some of the tokio specific code into one centralized location, and also abstracts the rest of the code on a runtime-agnostic codec type.

**Comment by TheButlah** [comment]
I've implemented the feature gating in #111

**Comment by mrdunky** [comment]
Would love to see this project succeed. 
However, I have read in a couple of places that webassembly does not have access to the the host environment when running in the browser. So no network APIs etc.
Did this change, or turn out to be incorrect? 

https://stackoverflow.com/questions/52102612/is-it-possible-to-submit-http-requests-with-webassembly

**Comment by TheButlah** [comment]
@mrdunky web assembly does not provide any networking apis in the spec, but because we are running in the browser we can call the browser's networking APIs via `wasm-bindgen` and `web-sys`:
https://rustwasm.github.io/wasm-bindgen/examples/fetch.html

So to answer what I think you're asking, yes we can basically do anything we could do in javascript in a frontend context.

**Comment by paddor** [comment]
For anyone still looking for alternative, [OMQ.ts](https://github.com/paddor/omq.ts) implements ZWS 2.0 over `ws://` and `wss://`, including `lz4+tcp://` compression transport ([benchmarks](https://github.com/paddor/omq.rs/blob/main/BENCHMARKS_COMPRESSION.md))
